File: app/src/main/java/com/example/pie/android/rest/RetrofitProvider.java
package com.example.pie.android.rest;

import com.example.pie.android.BuildConfig;
import com.google.gson.GsonBuilder;

import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitProvider {

    private static final String BASE_URL = BuildConfig.SERVER_URL;

    private static RetrofitProvider instance = new RetrofitProvider();

    private Retrofit retrofit;

    private RetrofitProvider() {
        retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(
                        new GsonBuilder()
                                .excludeFieldsWithoutExposeAnnotation()
                                .create()))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();
    }

    public static RetrofitProvider getInstance() {
        return instance;
    }

    public <T> T create(Class<T> service){
        return retrofit.create(service);
    }
}

File: app/src/main/java/com/example/pie/android/rest/RetrofitProvider.kt
package com.example.pie.android.rest

import com.example.pie.android.SERVER_URL
import com.google.gson.GsonBuilder

import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

class RetrofitProvider private constructor() {

    private val retrofit: Retrofit = Retrofit.Builder()
                .baseUrl(SERVER_URL)
                .addConverterFactory(GsonConverterFactory.create(
                        GsonBuilder()
                                .excludeFieldsWithoutExposeAnnotation()
                                .create()))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build()

    fun <T> create(service: Class<T>): T {
        return retrofit.create(service)
    }

    companion object {

        val instance by lazy { RetrofitProvider() }
    }
}
